Demo/Directory_watcher/dr_watcher_test_1.cpp

```diff
@@ -0,0 +1,81 @@
+#include <iostream>
+#include <thread>
+#include <chrono>
+#include <fstream>
+#include "dr_watcher.h"
+
+using namespace std::chrono;
+
+// Helper function to generate file events
+void generate_file_events(const std::string& file_path) {
+    std::this_thread::sleep_for(milliseconds(100)); // Allow watcher to start
+    std::ofstream file(file_path, std::ios::app);
+    file << "Writing to file\n"; // Trigger IN_MODIFY
+    file.close();
+    std::ifstream access_file(file_path); // Trigger IN_ACCESS
+    access_file.close();
+    std::remove(file_path.c_str()); // Trigger IN_DELETE_SELF
+}
+
+// Helper function to generate directory events
+void generate_directory_events(const std::string& dir_path) {
+    std::this_thread::sleep_for(milliseconds(100)); // Allow watcher to start
+    std::string file_path = dir_path + "/test_file.txt";
+    std::ofstream file(file_path); // Trigger IN_CREATE
+    file.close();
+    std::ofstream mod_file(file_path, std::ios::app);
+    mod_file << "Modifying file\n"; // Trigger IN_MODIFY
+    mod_file.close();
+    std::remove(file_path.c_str()); // Trigger IN_DELETE
+}
+
+// Benchmark monitoring a file
+void benchmark_file_watching() {
+    const std::string file_path = "test_file.txt";
+    std::ofstream file(file_path); // Create a file
+    file.close();
+
+    auto start = high_resolution_clock::now();
+    DirectoryWatcher watcher(file_path);
+
+    std::thread event_thread(generate_file_events, file_path);
```

Comment:
Remove anything connected with threads.

Reply:
I recommend testing it like this.
1. Write a bash script1 which sets up the necessary infrastructure.
2. Write a bash script2 which makes changes to that infrastructure.
3. Write a C/C++ program watcher_demo which takes a path as an argument and "watches" over it and logs any action made to the path into the stdout.
4. Write a bash script3 which runs script1 to set the infrastructure up, runs directory_demo in the background and redirects stdout to "results.txt" to watch over that infrastructure, runs script2 to make some changes to that infrastructure (which will be visible to the watcher in directory_demo program) and then compares the "results.txt" with the expected values.